File: packages/wasm-as/assembly/msgpack/Write.ts
import { Nullable } from "./Nullable";

export interface Write {
  writeNil(): void;
  writeBool(value: bool): void;
  writeInt8(value: i8): void;
  writeInt16(value: i16): void;
  writeInt32(value: i32): void;
  writeInt64(value: i64): void;
  writeUInt8(value: u8): void;
  writeUInt16(value: u16): void;
  writeUInt32(value: u32): void;
  writeUInt64(value: u64): void;
  writeFloat32(value: f32): void;
  writeFloat64(value: f64): void;
  writeStringLength(length: u32): void;
  writeString(value: string): void;
  writeBytesLength(length: u32): void;
  writeBytes(ab: ArrayBuffer): void;
  writeArrayLength(length: u32): void;
  writeArray<T>(
    a: Array<T>,
    fn: (writer: Write, item: T) => void
  ): void;
  writeMapLength(length: u32): void;
  writeMap<K, V>(
    m: Map<K, V>,
    keyFn: (writer: Write, key: K) => void,
    valueFn: (writer: Write, value: V) => void
  ): void;

  writeNullableBool(value: Nullable<bool>): void;
  writeNullableInt8(value: Nullable<i8>): void;
  writeNullableInt16(value: Nullable<i16>): void;
  writeNullableInt32(value: Nullable<i32>): void;
  writeNullableInt64(value: Nullable<i64>): void;
  writeNullableUInt8(value: Nullable<u8>): void;
  writeNullableUInt16(value: Nullable<u16>): void;
  writeNullableUInt32(value: Nullable<u32>): void;
  writeNullableUInt64(value: Nullable<u64>): void;
  writeNullableFloat32(value: Nullable<f32>): void;
  writeNullableFloat64(value: Nullable<f64>): void;
  writeNullableString(value: string | null): void;
  writeNullableBytes(ab: ArrayBuffer | null): void;
  writeNullableArray<T>(
    a: Array<T> | null,
    fn: (writer: Write, item: T) => void
  ): void;
  writeNullableMap<K, V>(
    m: Map<K, V> | null,
    keyFn: (writer: Write, key: K) => void,
    valueFn: (writer: Write, value: V) => void
  ): void;
}

File: packages/wasm/no-std-rs/src/msgpack/write.rs
use crate::Context;
use alloc::{collections::BTreeMap, string::String, vec::Vec};
use core::hash::Hash;
use num_bigint::BigInt;

pub trait Write: Clone {
    fn write_nil(&mut self);
    fn write_bool(&mut self, value: bool);
    fn write_i8(&mut self, value: i8);
    fn write_i16(&mut self, value: i16);
    fn write_i32(&mut self, value: i32);
    fn write_i64(&mut self, value: i64);
    fn write_u8(&mut self, value: u8);
    fn write_u16(&mut self, value: u16);
    fn write_u32(&mut self, value: u32);
    fn write_u64(&mut self, value: u64);
    fn write_f32(&mut self, value: f32);
    fn write_f64(&mut self, value: f64);
    fn write_string_length(&mut self, length: u32);
    fn write_string(&mut self, value: &String);
    fn write_str(&mut self, value: &str);
    fn write_bytes_length(&mut self, length: u32);
    fn write_bytes(&mut self, buf: &[u8]);
    fn write_bigint(&mut self, value: &BigInt);
    fn write_array_length(&mut self, length: u32);
    fn write_array<T: Clone>(&mut self, a: &[T], arr_fn: impl FnMut(&mut Self, &T));
    fn write_map_length(&mut self, length: u32);
    fn write_map<K, V: Clone>(
        &mut self,
        map: &BTreeMap<K, V>,
        key_fn: impl FnMut(&mut Self, &K),
        val_fn: impl FnMut(&mut Self, &V),
    ) where
        K: Clone + Eq + Hash + Ord;
    fn write_nullable_bool(&mut self, value: &Option<bool>);
    fn write_nullable_i8(&mut self, value: &Option<i8>);
    fn write_nullable_i16(&mut self, value: &Option<i16>);
    fn write_nullable_i32(&mut self, value: &Option<i32>);
    fn write_nullable_i64(&mut self, value: &Option<i64>);
    fn write_nullable_u8(&mut self, value: &Option<u8>);
    fn write_nullable_u16(&mut self, value: &Option<u16>);
    fn write_nullable_u32(&mut self, value: &Option<u32>);
    fn write_nullable_u64(&mut self, value: &Option<u64>);
    fn write_nullable_f32(&mut self, value: &Option<f32>);
    fn write_nullable_f64(&mut self, value: &Option<f64>);
    fn write_nullable_string(&mut self, value: &Option<String>);
    fn write_nullable_bytes(&mut self, buf: &Option<Vec<u8>>);
    fn write_nullable_bigint(&mut self, value: &Option<BigInt>);
    fn write_nullable_array<T: Clone>(
        &mut self,
        a: &Option<Vec<T>>,
        arr_fn: impl FnMut(&mut Self, &T),
    );
    fn write_nullable_map<K, V: Clone>(
        &mut self,
        map: &Option<BTreeMap<K, V>>,
        key_fn: impl FnMut(&mut Self, &K),
        val_fn: impl FnMut(&mut Self, &V),
    ) where
        K: Clone + Eq + Hash + Ord;
    fn context(&mut self) -> &Context;
}
